File: apps/android/app/src/main/java/dev/flextranslate/audio/AudioPipeline.kt
package dev.flextranslate.audio

import dev.flextranslate.foundation.AsrProvider
import dev.flextranslate.foundation.AudioFrame
import dev.flextranslate.foundation.TranscriptEvent
import java.util.ArrayDeque

/**
 * Связка между сырыми кадрами с микрофона и стеком распознавания.
 *
 * На каждый принятый [AudioFrame] конвейер:
 *  1. отдаёт кадр в подставленный [AsrProvider] и собирает [TranscriptEvent]'ы,
 *  2. кладёт кадр в кольцевой буфер ограниченного размера с backpressure «выкидываем самый старый»
 *     (медленный/залипший потребитель не раздувает память — вместо этого теряется старейший кадр),
 *  3. прогоняет подставленный [Vad] и запоминает последние [VadState] и [VadEvent],
 *  4. публикует [TranscriptEvent]'ы и снимок состояния в необязательный наблюдатель [onUpdate].
 *
 * Потоки: [accept] вызывается ОДНИМ потоком захвата (распознаватель держит своё состояние стрима и
 * не реентерабелен), поэтому вызов [AsrProvider.accept] из шага 1 идёт ВНЕ [lock] — лок защищает
 * только общие поля VAD/буфера/снимка, чтобы параллельный читатель (например, [currentVadState] из
 * UI-потока) видел согласованный снимок. Наблюдатель тоже зовём вне лока — иначе словим реентрантный
 * дедлок.
 */
class AudioPipeline(
    private val asrProvider: AsrProvider,
    private val vad: Vad,
    private val ringCapacity: Int = DEFAULT_RING_CAPACITY,
    private val onUpdate: (Snapshot) -> Unit = {},
) {

    /** Неизменяемый снимок конвейера после обработки кадра. */
    data class Snapshot(
        val vadState: VadState,
        val latestEvent: VadEvent?,
        val transcripts: List<TranscriptEvent>,
        val bufferDepth: Int,
        val droppedFrames: Long,
    )

    private val lock = Any()
    private val ring = ArrayDeque<AudioFrame>(ringCapacity)
    private var vadState: VadState = VadState.SILENCE
    private var latestEvent: VadEvent? = null
    private var droppedFrames: Long = 0L

    init {
        require(ringCapacity > 0) { "ringCapacity must be positive, was $ringCapacity" }
    }

    val currentVadState: VadState get() = synchronized(lock) { vadState }
    val latestVadEvent: VadEvent? get() = synchronized(lock) { latestEvent }
    val bufferDepth: Int get() = synchronized(lock) { ring.size }
    val totalDroppedFrames: Long get() = synchronized(lock) { droppedFrames }

    /** Обрабатывает один захваченный кадр. Зовётся из потока захвата. */
    fun accept(frame: AudioFrame) {
        val transcripts = asrProvider.accept(frame)
        val snapshot = synchronized(lock) {
            enqueueDropOldest(frame)
            val event = vad.accept(frame)
            if (event != null) {
                latestEvent = event
                vadState = when (event) {
                    is VadEvent.SpeechStart -> VadState.SPEECH
                    is VadEvent.SpeechEnd -> VadState.SILENCE
                }
            }
            Snapshot(
                vadState = vadState,
                latestEvent = latestEvent,
                transcripts = transcripts,
                bufferDepth = ring.size,
                droppedFrames = droppedFrames,
            )
        }
        onUpdate(snapshot)
    }

    /** Сбрасывает VAD, чистит буфер и ресетит ASR-провайдер. Захват сначала надо остановить. */
    fun reset() {
        synchronized(lock) {
            ring.clear()
            vad.reset()
            vadState = VadState.SILENCE
            latestEvent = null
            droppedFrames = 0L
        }
        asrProvider.reset()
    }

    /** Копия буферизованных кадров в порядке поступления (сначала старые). Для тестов/инспекции. */
    fun drainBufferedFrames(): List<AudioFrame> = synchronized(lock) { ring.toList() }

    private fun enqueueDropOldest(frame: AudioFrame) {
        if (ring.size >= ringCapacity) {
            ring.pollFirst()
            droppedFrames += 1
        }
        ring.addLast(frame)
    }

    companion object {
        /** ~5 с аудио 16 кГц при кадрах по 320 сэмплов (20 мс) — запас есть, память ограничена. */
        const val DEFAULT_RING_CAPACITY: Int = 250
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/Audio/AudioPipeline.swift
import Foundation

// Аудио-конвейер (WS2): стык между сырым захватом и слоем ASR/VAD.
//
// На каждый кадр: (1) кладёт его в кольцевой буфер с вытеснением старого и фикс-размером —
// чтобы медленный потребитель не разнёс память, (2) гоняет переданный VAD и запоминает
// последнее состояние/событие, (3) передаёт кадр в переданный `AsrProvider`. Сейчас ASR —
// это заглушка A1 (возвращает []), так что транскрипт не выдумывается — наружу идёт только
// реальное состояние VAD.
//
// Намеренно не потокобезопасен: тап захвата зовёт `accept` последовательно из своего
// колбэк-потока, а `LiveSessionModel` уводит апдейты состояния на @MainActor. Своих
// примитивов синхронизации у конвейера нет.
final class AudioPipeline {
    // Итог обработки одного кадра — чтобы вызывающий код и телеметрия могли среагировать
    // на переход VAD и на давление в буфере, не перечитывая состояние.
    struct FrameOutcome: Equatable {
        let vadEvent: VadEvent?
        let droppedOldest: Bool
    }

    private let vad: Vad
    private let asr: AsrProvider
    private let capacity: Int

    // Кольцевой буфер последних кадров с вытеснением старого. Ограничен `capacity`; когда
    // полон, перед добавлением нового выкидывается самый старый.
    private var ringBuffer: [AudioFrame] = []
    private var head = 0

    private(set) var vadState: VadState = .silence
    private(set) var latestEvent: VadEvent?

    // Последние события транскрипта от ASR (на этапе A1 пусто).
    private(set) var transcript: [TranscriptEvent] = []

    init(vad: Vad, asr: AsrProvider, capacity: Int = 64) {
        self.vad = vad
        self.asr = asr
        self.capacity = max(1, capacity)
        ringBuffer.reserveCapacity(self.capacity)
    }

    var bufferDepth: Int { ringBuffer.count }
    var speechActive: Bool { vadState == .speech }

    // Прогоняет один захваченный кадр через буфер -> VAD -> ASR. Возвращает итог
    // (переход VAD, выкинули ли старый кадр).
    @discardableResult
    func accept(_ frame: AudioFrame) -> FrameOutcome {
        let dropped = enqueue(frame)

        let event = vad.accept(frame)
        if let event {
            latestEvent = event
            switch event {
            case .speechStart:
                vadState = .speech
            case .speechEnd:
                vadState = .silence
            }
        }

        let events = asr.accept(frame: frame)
        if !events.isEmpty {
            transcript = events
        }

        return FrameOutcome(vadEvent: event, droppedOldest: dropped)
    }

    // Сброс буфера, VAD и ASR под новую сессию захвата.
    func reset() {
        ringBuffer.removeAll(keepingCapacity: true)
        head = 0
        vadState = .silence
        latestEvent = nil
        transcript = []
        vad.reset()
        asr.reset()
    }

    // Добавление с ограничением. Возвращает true, если ради места вытеснили самый старый
    // кадр (сигнал back-pressure). Двигаем head-индекс, чтобы в установившемся режиме
    // добавление оставалось O(1), когда буфер уже полон.
    private func enqueue(_ frame: AudioFrame) -> Bool {
        if ringBuffer.count < capacity {
            ringBuffer.append(frame)
            return false
        }
        ringBuffer[head] = frame
        head = (head + 1) % capacity
        return true
    }
}
